content/docs/rfcs/35/README.md

```diff
@@ -0,0 +1,336 @@
+; protocol ID
+protocol-id                 = 1OCTET
+
+; contains the size of handshake-message
+handshake-message-len       = 1OCTET
+
+; contains one or more Diffie-Hellman public keys
+handshake-message           = *OCTET
+
+; contains the size of message-len
+transport-message-len-len   = 1OCTET
+
+; contains the size of transport-message
+transport-message-len       = *OCTET
+
+; contains the transport message, eventually encrypted
+transport-message           = *OCTET
+
+; contains authentication data for transport-message, if encrypted
+transport-message-auth        = 16OCTET
+
+; the Waku WakuMessage payload field
+payload  = protocol-id handshake-message-len handshake-message transport-message-len-len transport-message-len transport-message transport-message-auth 
+```
+
+### Protocol Payload Format
+
+Based on the specified `protocol-id`, 
+the Waku message `payload` field will encode different types of protocol-dependent messages. 
+
+In particular, if `protocol-id` is
+- `0`:  payload encodes an [after-handshake](#After-handshake) message.  
+    - `handshake-message-len` MAY be 0;
+    -  `transport-message` contains the Noise transport message;
+- `10`,`11`,`12`,`13`: payload encodes a supported Noise handshake message.  
+    -  `transport-message` contains the Noise transport message;
+- `30`: payload encapsulate a `ChaChaPoly` ciphertext `ct`.  
+    - `handshake-message-len` is set to `0`;
+    - `transport-message` contains the concatenation of the encryption nonce (12 bytes) followed by the ciphertext `ct`;
+    - `transport-message-len-len` and `transport-message-len` are set accordingly to `transport-message` length;
+    - `transport-message-auth` contains the authentication data for `ct`.
+
+
+### Public Keys Serialization
+
+Diffie-Hellman public keys can be trasmitted in clear 
+or in encrypted form with authentication data attached. 
+To distinguish between these two cases, public keys are serialized as the concatenation of the following three fields:
+
+- `flag`: 
+is equal to `1` if the public key is encrypted; 
+`0` otherwise (1 byte);
+- `pk`: 
+if `flag = 0`, it contains an encoding of the X coordinate of the public key. 
+If `flag = 1`, it contains a symmetric encryption of an encoding of the X coordinate of the public key;
+- `pk-auth`: 
+if `flag = 0`, it is empty; 
+if `flag = 1`, it contains the authentication data for `pk`;
+
+The corresponding serialization is obtained as `flag pk pk-auth`.
+
+As regards the underlying supported [cryptographic primitives](#Cryptographic-primitives):
+- `Curve25519` public keys X coordinates are encoded in little-endian as 32 bytes arrays;
+- `ChaChaPoly` authentication data consists of 16 bytes 
+(nonces are implicitely defined by Noise [processing rules](http://www.noiseprotocol.org/noise.html#processing-rules)).
+
+In all supported Noise protocols, 
+parties' static keys are transmitted encrypted, 
+while ephemeral keys MAY be encrypted after a handshake is complete.
```

Comment:
If this is the case i.e., both ephemeral pks and static pks can be encrypted, how would you distinguish between these two in the current serialization pattern?
btw,  which handshake pattern contains an encrypted ephemeral key? with a quick check I could not find

Reply:
During an handshake, you don't need to distinguish, since the handshake pattern is pre-agreed (so you know the sequence of ephemeral-static encrypted/unencrypted keys). After handshake you'll only have ephemeral keys exchanged to refresh with new randomness encryption keys. The distinction between encrypted/unencrypted can be done simply looking at the authentication data.

Reply:
If you know the order of pks by knowing the handshake pattern, then why do we need a flag to indicate encryption? isn't it also clear from the pattern?

Reply:
Because in a handshake message you may have a concatenation of multiple public keys. Without a flag you may confuse authentication data with the next public key, or waste bytes for empty authentication data in case you keep sizes fixed.

Reply:
> you may confuse authentication data with the next public key

Can you please give an example of such a situation? not numerically just symbolically

Reply:
It is not impossible, only less structured. 

For example, without flags
`handskake-message = pk1 pk1-auth pk2 pk3` 
You read the first 32 bytes and you get `pk1`. You read next 16 bytes, if authenticating `pk1`, decrypt `pk1`. Otherwise continue reading next 16 bytes and store the last 32 bytes read and store as `pk2`, etc.

Probably hard to do, but you may generate a public key with MSB set to authentication data for previous key and confuse the implementation, although the handshake will simply fail.

Reply:
Thanks for the explanation, isn't the presence of `pk1_auth` dictated by the handshake pattern? i.e., if there has been a shared key established prior to `pk1`, then an encrypted `pk1` is expected? which means the next item in the handshake message should be `pk1-auth`?

Reply:
If this comment of mine also requires further discussion, let's issue it up and resume our conversion there.

Reply:
I would say both yes and no. The handshake is pre-agreed, so everyone knows what to expect. In after-handshake, although you may exchange only ephemeral keys for which there is no need to use encryption, you may encrypt them as well (for example to authenticate the ephemeral key itself, but again encryption will fail if you're cheating). I think the flag may be useful in the future for Noise pipes, i.e. while executing multiple handshakes in a row: you can indeed upgrade the whole handshake to a fully encrypted version, although you can then use a `psk` version of the handshake then.
It is also true that besides handshakes, I don't expect to exchange more than one public key during normal communication, so checking the length of `handshake-message` would suffice to understand if the public key is encrypted or not, and we can in fact remove the `flag`.